### PR TITLE
Set Up CloudWatch Alarms for CPU Utilization

### PR DESCRIPTION
**Pull request by ilan-gold**

### CODE REVIEW

**Comment by ilan-gold** [comment]
@hackdna this has been tested

File: deployment/terraform/modules/ec2/main.tf [hunk 1]
```diff
@@ -347,3 +347,21 @@ resource "aws_elb" "https" {
     unhealthy_threshold = 4
   }
 }
+
+resource "aws_cloudwatch_metric_alarm" "cpu_utilization" {
```
Comment (hackdna) [inline deployment/terraform/modules/ec2/main.tf]: `resource "aws_cloudwatch_metric_alarm" "app_server_cpu_utilization"`
We may have other CPU utilization alarms.

File: deployment/terraform/modules/ec2/main.tf [hunk 2]
```diff
@@ -347,3 +347,21 @@ resource "aws_elb" "https" {
...
+  alarm_name                = "cpu-utilization-limit-exceeded"
```
Comment (hackdna) [inline deployment/terraform/modules/ec2/main.tf]: Alarm name must be unique within AWS account, and also should be descriptive.
`alarm_name = "${var.resource_name_prefix}-app-server-cpu-utilization"`

File: deployment/terraform/modules/ec2/main.tf [hunk 3]
```diff
@@ -347,3 +347,21 @@ resource "aws_elb" "https" {
...
+  comparison_operator       = "GreaterThanOrEqualToThreshold"
+  evaluation_periods        = "5"
+  metric_name               = "CPUUtilization"
+  namespace                 = "AWS/EC2"
+  period                    = "300"
```
Comment (hackdna) [inline deployment/terraform/modules/ec2/main.tf]: If I'm not mistaken this is going to raise an alarm if CPU utilization has been >=25% for over 25 minutes. Is that what we want?

File: deployment/terraform/modules/ec2/main.tf [hunk 4]
```diff
@@ -347,3 +347,21 @@ resource "aws_elb" "https" {
...
+  statistic                 = "Average"
+  threshold                 = "25"
+  alarm_description         = "This metric monitors ec2 cpu utilization"
```
Comment (hackdna) [inline deployment/terraform/modules/ec2/main.tf]: `"Monitors CPU utilization of ${aws_instance.app_server.tags.Name}"`

File: deployment/terraform/modules/ec2/main.tf [hunk 5]
```diff
@@ -347,3 +347,21 @@ resource "aws_elb" "https" {
     unhealthy_threshold = 4
   }
 }
+
+resource "aws_cloudwatch_metric_alarm" "cpu_utilization" {
+  alarm_name                = "cpu-utilization-limit-exceeded"
+  comparison_operator       = "GreaterThanOrEqualToThreshold"
```
Comment (hackdna) [inline deployment/terraform/modules/ec2/main.tf]: Terraform style guide calls for aligning all equal signs within a block to the longest attribute name which is this one.

File: deployment/terraform/modules/ec2/main.tf (hunk 3 above)
Comment (ilan-gold) [reply deployment/terraform/modules/ec2/main.tf]: This is a judgement call as far as I can tell.  25 minutes seemed like a good amount of time to allow for spikes/prolonged periods of usage.

Comment (hackdna) [reply deployment/terraform/modules/ec2/main.tf]: I see. Also, how and to whom this alarm notification will be delivered?

Comment (ilan-gold) [reply deployment/terraform/modules/ec2/main.tf]: I was under the impression that a notification would go to the root account user, but that was clearly mistaken.  Not sure how I came to that conclusion.  In any case: https://www.terraform.io/docs/providers/aws/r/sns_topic_subscription.html#email

In short, setting up an email endpoint for an sns topic entirely within Terraform is impossible.  My vote would be to create the topic/subscription outside Terraform within AWS and just pass in the arn as a default option to the alarms for subscription.

Comment (hackdna) [reply deployment/terraform/modules/ec2/main.tf]: Yes, email subscriptions require an out-of-band confirmation. Also, some deployments may need a different type of notification or no notification at all. Could you set it up such that it is possible to set up a notification but it is not required by default?

Comment (ilan-gold) [reply deployment/terraform/modules/ec2/main.tf]: Yes I think that makes sense

File: deployment/terraform/modules/rds/main.tf [hunk 6]
```diff
@@ -45,3 +45,20 @@ resource "aws_db_instance" "default" {
...
+
+resource "aws_cloudwatch_metric_alarm" "default_rds_cpu_utilization" {
+  alarm_name          = "${var.resource_name_prefix}-default-rds-cpu-utilization"
+  comparison_operator = "GreaterThanOrEqualToThreshold"
+  evaluation_periods  = "5"
+  metric_name         = "CPUUtilization"
+  namespace           = "AWS/RDS"
+  period              = "300"
+  statistic           = "Average"
+  threshold           = "25"
```
Comment (hackdna) [inline deployment/terraform/modules/rds/main.tf]: Given the data (https://github.com/refinery-platform/refinery-platform/issues/2965#issuecomment-438707280) I'd say a 10% threshold would be more appropriate.

File: deployment/terraform/modules/rds/main.tf [hunk 7]
```diff
@@ -45,3 +45,20 @@ resource "aws_db_instance" "default" {
...
+  alarm_description   = "Monitors CPU utilization of default RDS instance"
```
Comment (hackdna) [inline deployment/terraform/modules/rds/main.tf]: `"Monitors CPU utilization of ${aws_db_instance.default.id}"` or whatever is the human readable name.

Comment (ilan-gold) [reply deployment/terraform/modules/rds/main.tf]: I think the most human readable name is "default."  There are no tags or the like for RDS.  We can change that though.

File: deployment/terraform/modules/rds/main.tf (hunk 6 above)
Comment (ilan-gold) [reply deployment/terraform/modules/rds/main.tf]: Ok

Comment (ilan-gold) [reply deployment/terraform/modules/rds/main.tf]: does this mean for both ec2 and rds?

File: deployment/terraform/live/vars.tf [hunk 8]
```diff
@@ -195,3 +195,8 @@ variable "data_volume_snapshot_id" {
   description = "A snapshot to base the EBS data volume off of"
   default     = ""
 }
+
+variable "alarm_sns_arn" {
+  description = "The arn of an sns topic to publish cloudwatch alarm metrics"
```
Comment (hackdna) [inline deployment/terraform/live/vars.tf]: "The ARN of an SNS topic to publish CloudWatch alarms"

**Pull request merged**